### PR TITLE
Package for running periodic tasks

### PR DESCRIPTION
**Pull request by jonasfj**

### CODE REVIEW

File: neat_periodic_task/test/neat_periodic_task_test.dart [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      return true;
+    }
+    return false;
+  }
+}
+
+void main() {
+  // Setup logging
+  Logger.root.level = Level.ALL;
+  Logger.root.onRecord.listen((r) => print(r));
```
Comment (sigurdm) [inline neat_periodic_task/test/neat_periodic_task_test.dart]: You could collect the loggings and do expectations on them

File: neat_periodic_task/test/neat_periodic_task_test.dart [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  test('schedule periodic task', () async {
+    int count = 0;
+    final scheduler = NeatPeriodicTaskScheduler(
+      name: 'test-task',
+      interval: Duration(milliseconds: 500),
+      timeout: Duration(milliseconds: 50),
+      minCycle: Duration(milliseconds: 100),
+      maxCycle: Duration(milliseconds: 250),
+      task: () async {
+        print('doing operation');
```
Comment (sigurdm) [inline neat_periodic_task/test/neat_periodic_task_test.dart]: Don't print in the test

File: neat_periodic_task/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+To be alerted when a task consistently fails to run, it is sufficient to alert
+if the log message `'### [ALIVE] neat-periodic-task: "<name>"'` is absent from
+logs for more than 24 hours. This message is printed when the status is checked
+and the scheduler decided not to run the task because it was completely
+successfully less than `interval` time ago.
+
+A simple robust monitoring strategy is to alert if the message
+`'### [ALIVE] neat-periodic-task: "<name>"'` have **not** occurred in 24 hours.
+This works regardless of the `interval` length, and will fire alerts even if
+server fails to boot.
```
Comment (sigurdm) [inline neat_periodic_task/README.md]: nit: even if the server

File: neat_periodic_task/lib/neat_periodic_task.dart [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    }
+
+    // Find time elapsed since last status update.
+    final now = DateTime.now().toUtc();
+    final elapsed = now.difference(status.updated);
+    _log.finest(
+        () => 'time elapsed since status for "$_name" was written $elapsed');
+
+    // If state is 'finished' the delay before next run is _interval, otherwise
+    // we assume state us 'running' as delay is only _timeout.
```
Comment (sigurdm) [inline neat_periodic_task/lib/neat_periodic_task.dart]: nit: us -> is

File: neat_periodic_task/lib/neat_periodic_task.dart [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+          '### [FAILED] neat-periodic-task: "$_name", timeout exceeded!');
+      return;
+    } catch (e, st) {
+      _log.shout('### [FAILED] neat-periodic-task: "$_name"', e, st);
+      return;
+    }
+
+    _log.finest(() => 'Attempting to set finished status for "$_name"');
+    final st = status.update(state: 'finished').serialize();
+    if (!await _statusProvider.set(st)) {
```
Comment (sigurdm) [inline neat_periodic_task/lib/neat_periodic_task.dart]: Should you update updated?

File: neat_periodic_task/lib/neat_periodic_task.dart [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    _log.finest(() => 'Attempt to claim task status for "$_name"');
+    final claimed = await _statusProvider.set(status.serialize());
+    if (!claimed) {
+      _log.finest(() => 'Failed to claim task status for "$_name"');
+      return;
+    }
+
+    try {
+      _log.info('### [START] neat-periodic-task: "$_name"');
+      await _task().timeout(_timeout);
```
Comment (sigurdm) [inline neat_periodic_task/lib/neat_periodic_task.dart]: You could upgrade the Task return-type with a cancel.

File: neat_periodic_task/lib/neat_periodic_task.dart [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+  Future<void> trigger() async {
+    _log.finest(() => 'NeatPeriodicTaskScheduler "$_name" trigger() called');
+    final status = await _getStatus();
+
+    // Find time elapsed since last status update.
+    final now = DateTime.now().toUtc();
+    final elapsed = now.difference(status.updated);
+
+    // If not running, or timed-out we run the task again.
+    if (status.state != 'running' || elapsed > _timeout) {
```
Comment (sigurdm) [inline neat_periodic_task/lib/neat_periodic_task.dart]: Perhaps extract the status strings into constants (or enums).

File: neat_periodic_task/lib/neat_periodic_task.dart [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+  Future<bool> set(List<int> status) async {
+    _value = status;
+    return true;
+  }
+}
+
+class _NeatStatusProviderWithRetry extends NeatStatusProvider {
+  final NeatStatusProvider _provider;
+  final RetryOptions _r;
+  _NeatStatusProviderWithRetry(this._provider, this._r);
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply neat_periodic_task/lib/neat_periodic_task.dart]: The type is private, so I'm not going to care much about it..

File: neat_periodic_task/lib/neat_periodic_task.dart [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+  /// [NeatStatusProvider].
+  ///
+  /// The [status] provider is effectively a light weight locking mechanism to
+  /// prevent the task from running concurrently. However, this only holds if
+  /// the task always completes or fails within [timeout], as it may otherwise
+  /// be restarted.
+  ///
+  /// If the task fails consistently, it will be retried at [timeout] delay,
+  /// this will continue indefinitely. Thus, it is sensible to pick a high
+  /// [timeout], if the operation is expensive and this can be tolerated.
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply neat_periodic_task/lib/neat_periodic_task.dart]: Yes, that's left as a wish list item for the future.

File: neat_periodic_task/lib/neat_periodic_task.dart (hunk 7 above)
Comment (jonasfj) [reply neat_periodic_task/lib/neat_periodic_task.dart]: Yeah, in an ideal world... on the other hand it just adds a lot of typing for no particular value...

File: neat_periodic_task/lib/neat_periodic_task.dart (hunk 5 above)
Comment (jonasfj) [reply neat_periodic_task/lib/neat_periodic_task.dart]: Let's instead change the name `updated` to `finished`, I think it's much better...

File: neat_periodic_task/lib/neat_periodic_task.dart (hunk 6 above)
Comment (jonasfj) [reply neat_periodic_task/lib/neat_periodic_task.dart]: Agree... it would be a neat feature.. But outside the scope for now.

**Pull request merged**